crates/chain/src/local_chain.rs

```diff
@@ -2,15 +2,166 @@
 
 use core::convert::Infallible;
 
-use alloc::collections::BTreeMap;
+use crate::collections::BTreeMap;
+use crate::{BlockId, ChainOracle};
+use alloc::sync::Arc;
 use bitcoin::BlockHash;
 
-use crate::{BlockId, ChainOracle};
+/// A structure that represents changes to [`LocalChain`].
+pub type ChangeSet = BTreeMap<u32, Option<BlockHash>>;
+
+/// A [`LocalChain`] checkpoint is used to find the agreement point between two chains and as a
+/// transaction anchor.
+///
+/// Each checkpoint contains the height and hash of a block ([`BlockId`]).
+///
+/// Internaly, checkpoints are nodes of a linked-list. This allows the caller to view the entire
+/// chain without holding a lock to [`LocalChain`].
+#[derive(Debug, Clone)]
+pub struct CheckPoint(Arc<CPInner>);
+
+/// The internal contents of [`CheckPoint`].
+#[derive(Debug, Clone)]
+struct CPInner {
+    /// Block id (hash and height).
+    block: BlockId,
+    /// Previous checkpoint (if any).
+    prev: Option<Arc<CPInner>>,
+}
```

Comment:
I'm wondering why you chose to create CPInner and CheckPoint instead of using LinkedList<BlockId>. Is it because you want to be able to hold a single CheckPoint?

Reply:
Of course, I'm just wondering why you're not using the std LinkedList: https://doc.rust-lang.org/std/collections/struct.LinkedList.html

Reply:
One requirement we have is that each node should be both `Sync + Send`. This way, the update thread can have to total view of the chain without keeping a lock on `LocalChain`.

Also, we only need a singly linked-list so `LinkedList` might be overkill.